src/macaron/build_spec_generator/cli_command_parser/gradle_cli_parser.py

```diff
@@ -0,0 +1,701 @@
+    """This option represents an optional negateable flag in Gradle CLI command.
+
+    For example: --build-cache/--no-build-cache
+    """
+
+    def is_valid_patch_option(self, patch: Any) -> TypeGuard[bool]:
+        """Return True if the provide patch value is compatible with the internal type of this option."""
+        return isinstance(patch, bool)
+
+    @staticmethod
+    def get_negated_long_name(long_name: str) -> str:
+        """Return the negated version of a long option name."""
+        return f"--no-{long_name.removeprefix('--')}"
+
+    def add_itself_to_arg_parser(self, arg_parse: argparse.ArgumentParser) -> None:
+        """Add a new argument to argparser.ArgumentParser representing this option."""
+        # We allow providing both the normal and negated form.
+        negated_long_name = self.get_negated_long_name(self.long_name)
+        dest = self.long_name.removeprefix("--").replace("-", "_")
+
+        # We set the default to None so that we don't print out these options
+        # if they are not provided in the original build command in to_cmd_tasks().
+        arg_parse.add_argument(
+            self.long_name,
+            action="store_true",
+            default=None,
+            dest=dest,
+        )
+
+        arg_parse.add_argument(
+            negated_long_name,
+            action="store_false",
+            default=None,
+            dest=dest,
+        )
+
+    def get_patch_type_str(self) -> str:
+        """Return the expected type for the patch value as string."""
+        return "bool"
+
+
+@dataclass
+class GradleSingleValue(OptionDef[str]):
+    """This option represents an option that takes a value in Grale CLI command."""
+
+    short_name: str | None
+
+    def is_valid_patch_option(self, patch: Any) -> TypeGuard[str]:
+        """Return True if the provide patch value is compatible with the internal type of this option."""
+        return isinstance(patch, str)
+
+    def add_itself_to_arg_parser(self, arg_parse: argparse.ArgumentParser) -> None:
+        """Add a new argument to argparser.ArgumentParser representing this option."""
+        if self.short_name:
+            arg_parse.add_argument(
+                *(self.short_name, self.long_name),
+            )
+        else:
+            arg_parse.add_argument(
+                self.long_name,
+            )
+
+    def get_patch_type_str(self) -> str:
+        """Return the expected type for the patch value as string."""
+        return "str"
+
+
+@dataclass
+class GradlePropeties(OptionDef[dict[str, str | None]]):
+    """This option represents an option used to define properties values of a Gradle CLI command.
+
+    This option can be defined multiple times and the values are appended into a list of string in argparse.
+    However, it's stored internally as a dictionary mapping between the system property name to its value.
+
+    In Gradle there are 2 options of this type:
+        - -D/--system-prop
+        - -P/--project-prop
```

Comment:
I think an example of how these are used would help distinguish this class from `GradleAppendedList`. Perhaps a mention of key value pairs or similar would also be of use.

Reply:
Definitely, I will add the examples.